lib/resty/wasmx/shm.lua

```diff
@@ -295,78 +320,186 @@ local function set_kv_value(zone, key, value, cas)
         error("value must be a string", 2)
     end
 
-    if type(cas) ~= "number" then
+    if cas == nil then
+        cas = 0
+
+    elseif type(cas) ~= "number" then
         error("cas must be a number", 2)
     end
 
     local shm = zone[WASM_SHM_KEY]
     local cname = ffi_new("ngx_str_t", { data = key, len = #key })
     local cvalue = ffi_new("ngx_str_t", { data = value, len = #value })
-    local written = ffi_new("uint32_t [1]")
+    local written = ffi_new("ngx_int_t[1]")
 
-    local rc = C.ngx_wa_ffi_shm_set_kv_value(shm, cname, cvalue, cas, written)
-    if rc == FFI_ERR then
-        return nil, "failed setting kv value, no memory"
+    local rc = C.ngx_wa_ffi_shm_kv_set(shm, cname, cvalue, cas, written)
+    if rc == FFI_ERROR then
+        return nil, "no memory"
+    end
+
+    if rc ~= FFI_OK then
+        return nil, "unknown error"
     end
 
     return tonumber(written[0])
 end
 
 
+local function metrics_define(zone, name, metric_type)
+    if type(name) ~= "string" or name == "" then
+        error("name must be a non-empty string", 2)
+    end
+
+    if not _metric_type_set[metric_type] then
+        local err = "metric_type must be one of" ..
+                    " resty.wasmx.shm.metrics.COUNTER," ..
+                    " resty.wasmx.shm.metrics.GAUGE, or" ..
+                    " resty.wasmx.shm.metrics.HISTOGRAM"
+
+        error(err, 2)
+    end
+
+    name = "lua." .. name
+
+    local cname = ffi_new("ngx_str_t", { data = name, len = #name })
+    local m_id = ffi_new("uint32_t [1]")
+
+    local rc = C.ngx_wa_ffi_shm_metric_define(cname, metric_type, m_id)
+    if rc == FFI_ERROR then
+        return nil, "no memory"
+    end
+
+    if rc == FFI_ABORT then
+        return nil, "name too long"
+    end
+
+    if rc ~= FFI_OK then
+        -- FFI_BUSY: unreachable
+        return nil, "unknown error"
+    end
+
```

Comment:
The check `rc ~= FFI_OK` above isn't really needed.

I noticed you changed `ngx_wa_metrics_define` to return `NGX_BUSY` if the function is called with an unknown metric type. I can't really see the benefit of discriminating the particular types of bad usage here. Specially considering that `ngx_wa_metrics_define` will never receive a bad metric type.

I think the existing error code convention is enough to capture errors that the user needs to be aware of, e.g. metric name too long.

Reply:
As written in the comment the code path is unreachable but the code is defensive and explicit.